tests/test_layers.py

```diff
@@ -115,6 +116,47 @@ def test_amplitude_db(amplitude, db):
                              atol=1e-5)
 
 
+@pytest.mark.parametrize('rate', [0.7, 0.9, 1.01, 1.3])
+@pytest.mark.parametrize('complex_specgrams', [
+    torch.randn(1, 2, 1025, 400, 2)
+])
+@pytest.mark.parametrize('hop_length', [256])
+def test_TimeStretch(complex_specgrams, rate, hop_length):
+
+    class use_double_precision:
+        def __enter__(self):
+            self.default_dtype = torch.get_default_dtype()
+            torch.set_default_dtype(torch.float64)
+
+        def __exit__(self, type, value, traceback):
+            torch.set_default_dtype(self.default_dtype)
+
+    with use_double_precision():
```

Comment:
For me, although the error is stacked with `cumsum`, with a lower precision it should be fine if the mean of precision error is zero. Could I ask you to double check like line-by-line if everything is fine?

Reply:
So if I compare the magnitudes with `torch.float32`, they're equal up to `atol=1e-4`. But the actual complex tensor is not (and you can see how along the time axis the results diverge more and more as the `cumsum` error accumulates). 

Do you mean check if the algorithm is fine? I'm like very sure that it is since simply changing to `torch.float64` makes librosa equal to ours > `atol=1e-10`. I just can't get the comparison to pass for default dtypes.

Reply:
![error](https://user-images.githubusercontent.com/12011058/59164163-5ec68a80-8abe-11e9-8411-5470fea87a7d.png)
Looking at the error diff there does seem to be something weird going on... but for sure numeric.

Reply:
![error_rates](https://user-images.githubusercontent.com/12011058/59164638-6ab54b00-8ac4-11e9-91a7-6ad29c967a5e.png)
Ok so this continues to be strange, can still see the cumulative error but it's smoother for either `rate > 1` or `rate < 1` with more than one decimal... (!?). Is some casting stuff going wrong? The problem is with `torch` since I've also compared the `torch.float32` vs. `torch.float64` and it produces similar patterns for rates 0.7, 0.8, etc. and not for 0.702, 0.81, 1.1, 1.2 type ones....

Reply:
Ha, that's interesting. Is this difference compared to librosa/float64 result? What happens if we compare our torch/float32 vs libosa/float32?

Reply:
> Is this difference compared to librosa/float64 result?

These last two plots were comparing `torch.float32` with librosa/float32. Like I said not too sure what's going on in the left, but the right one is a good representation of the pattern I see when not using float64. 

Even comparing our implementation for the same input, diff between `torch.float64` and `torch.float32`:
![error](https://user-images.githubusercontent.com/12011058/59165990-a1955c00-8ad8-11e9-883f-bb3ff9e6ce28.png)

And doing the same for librosa/float32 vs librosa/float64:
![error_librosa](https://user-images.githubusercontent.com/12011058/59166017-3bf59f80-8ad9-11e9-9a78-556368850b8d.png)

So this isn't something unique to our implementation (or to pytorch for that matter). 

We could just compare the magnitudes of the stretched spectrograms and set a higher tolerance like `1e-3`? But if we want to do straight comparison looks like we'll need double precision...

Reply:
Strange.. although librosa's implementation is frame-wise computation, there's no feedback from `d_stretch`, so it should be fine with doing it once like ours. I'm not sure if this should block the procedure. @faroit Do you have any idea? @ksanjeevan How does the result sound like? As a baseline, not very scientific but we can check out some long audio files and if they sound ok, probably we can merge at least for the Phase 1.

Reply:
> although librosa's implementation is frame-wise computation, there's no feedback from d_stretch

What do you mean? Our implementation is the same as librosa just doing the operations with tensors instead of a for loop. They also have an accumulated phase.

I don't think this should block the procedure. Our implementation produces the same result as librosa's when we both use float64, and  presents the same kind of numeric error as librosa when using float32. The reason they aren't equal for float32 I guess is the error behaves like a random walk and piles up in time (?). Not quite sure. 

But yeah :+1: to do some sanity checks before we merge in Phase 1!

Reply:
> instead of a for loop.

That's what I meant by 'frame-wise computation'

Ok good, tentatively, let's use the current test and see. We can worry later if `torch/audio` maintainers see it as a problem.